examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+      : fe_values(fe,
+                  QGauss<dim>(quadrature_degree),
+                  update_values | update_gradients | update_hessians |
+                    update_quadrature_points | update_JxW_values)
+    {}
+
+    ScratchData(const ScratchData<dim> &scratch_data)
+      : fe_values(scratch_data.fe_values.get_fe(),
+                  scratch_data.fe_values.get_quadrature(),
+                  update_values | update_gradients | update_hessians |
+                    update_quadrature_points | update_JxW_values)
+    {}
+
+    FEValues<dim> fe_values;
+  };
+  struct CopyData
+  {
+    unsigned int level;
+    unsigned int dofs_per_cell;
+
+    FullMatrix<double>                   cell_matrix;
+    Vector<double>                       cell_rhs;
+    std::vector<types::global_dof_index> local_dof_indices;
+  };
+
+
+  struct Settings
+  {
+    bool try_parse(const std::string &prm_filename);
+
+    unsigned int fe_degree;
+    std::string  smoother_type;
+    std::string  dof_renum;
+    bool         with_sd;
+    bool         output;
+  };
+
+
+  bool Settings::try_parse(const std::string &prm_filename)
+  {
+    ParameterHandler prm;
+
+    prm.declare_entry("fe degree",
+                      "1",
+                      Patterns::Integer(0),
+                      "Finite Element degree");
+    prm.declare_entry("smoother type",
+                      "block sor",
+                      Patterns::Selection("sor|jacobi|block sor|block jacobi"),
+                      "Smoother Type: sor|jacobi|block sor|block jacobi");
+    prm.declare_entry("dof renumbering",
+                      "downstream",
+                      Patterns::Selection("none|random|downstream|upstream"),
+                      "Dof renumbering: none|random|downstream|upstream");
+    prm.declare_entry("with sd",
+                      "true",
+                      Patterns::Bool(),
+                      "With streamline diffusion: true|false");
+    prm.declare_entry("output",
+                      "true",
+                      Patterns::Bool(),
+                      "Generate graphical output: true|false");
+
+    try
+      {
+        prm.parse_input(prm_filename);
+      }
+    catch (const dealii::PathSearch::ExcFileNotFound &)
+      {
+        if (prm_filename.size() > 0)
+          std::cerr << "ERRROR: could not open the .prm file '" << prm_filename
+                    << "'" << std::endl;
+        else
+          std::cerr << "Usage: please pass a .prm file as the first argument"
+                    << std::endl;
+
+        prm.print_parameters(std::cout, ParameterHandler::Text);
+        return false;
+      }
+    this->fe_degree     = prm.get_integer("fe degree");
+    this->smoother_type = prm.get("smoother type");
+    this->dof_renum     = prm.get("dof renumbering");
+    this->with_sd       = prm.get_bool("with sd");
+    this->output        = prm.get_bool("output");
+
+    return true;
```

Comment:
This return value is used in `main()` to decide whether or not to continue with the program. We want to catch this exception `dealii::PathSearch::ExcFileNotFound` so that we can print the correct parameters to the screen. Are you suggesting that I `return void` for this function, and if I catch `dealii::PathSearch::ExcFileNotFound`, then print the parameters and rethrow the exception?

Reply:
From a software design perspective, `return` statements are used to return things a function has computed. Exceptions are used for "exceptional" situations, and what you have here is exactly the kind. The way I'd treat what you have is by either just not catching the exception in the code above at all and letting it propagate to `main()` (which has the downside that `main()` will print an error message that contains less information than one might want) or, maybe better, by catching the exception as you do, but instead of printing the error message directly, put it into an `ExcMessage` object and throw that instead. In that case, `main()` will print the re-packaged error message, and you won't have to have the extra logic here with dealing with return values, nor the extra logic with asking for the returned value and making decisions based on it.

Does this make sense?